nav2_collision_monitor/src/polygon.cpp

```diff
@@ -388,51 +420,114 @@ bool Polygon::getParameters(
-          poly_.push_back(point);
-        }
-        first = !first;
-      }
-
       // Do not need to proceed further, if "points" parameter is defined.
       // Static polygon will be used.
-      return true;
+      return setPolygonShape(poly_row, poly_);
     } catch (const rclcpp::exceptions::ParameterUninitializedException &) {
       RCLCPP_INFO(
         logger_,
         "[%s]: Polygon points are not defined. Using dynamic subscription instead.",
         polygon_name_.c_str());
     }
 
-    if (action_type_ == STOP || action_type_ == SLOWDOWN || action_type_ == LIMIT) {
-      // Dynamic polygon will be used
+    try {
+      if (action_type_ == STOP || action_type_ == SLOWDOWN || action_type_ == LIMIT) {
+        // Dynamic polygon will be used
+        nav2_util::declare_parameter_if_not_declared(
+          node, polygon_name_ + ".polygon_sub_topic", rclcpp::PARAMETER_STRING);
+        polygon_sub_topic =
+          node->get_parameter(polygon_name_ + ".polygon_sub_topic").as_string();
+        return true;
+      } else if (action_type_ == APPROACH) {
+        // Obtain the footprint topic to make a footprint subscription for approach polygon
+        nav2_util::declare_parameter_if_not_declared(
+          node, polygon_name_ + ".footprint_topic", rclcpp::PARAMETER_STRING);
+        footprint_topic =
+          node->get_parameter(polygon_name_ + ".footprint_topic").as_string();
+        return true;
+      }
+    } catch (const rclcpp::exceptions::ParameterUninitializedException &) {
+      RCLCPP_INFO(
+        logger_,
+        "[%s]: Polygon topic are not defined. Using polygon velocity instead.",
+        polygon_name_.c_str());
+    }
+
+    // Polygon velocity will be used
+    nav2_util::declare_parameter_if_not_declared(
+      node, polygon_name_ + ".polygon_velocity", rclcpp::PARAMETER_STRING_ARRAY);
+    std::vector<std::string> polygon_velocity_array = node->get_parameter(
+      polygon_name_ + ".polygon_velocity").as_string_array();
+
+    for (std::string polygon_velocity_name : polygon_velocity_array) {
+
+      std::string polygon_name = polygon_velocity_name;
+
       nav2_util::declare_parameter_if_not_declared(
-        node, polygon_name_ + ".polygon_sub_topic", rclcpp::PARAMETER_STRING);
-      polygon_sub_topic =
-        node->get_parameter(polygon_name_ + ".polygon_sub_topic").as_string();
-    } else if (action_type_ == APPROACH) {
-      // Obtain the footprint topic to make a footprint subscription for approach polygon
+        node, polygon_name_ + "." + polygon_velocity_name + ".points",
+        rclcpp::PARAMETER_DOUBLE_ARRAY);
+      std::vector<double> polygon_points = node->get_parameter(
+        polygon_name_ + "." + polygon_velocity_name + ".points").as_double_array();
+
+      std::vector<Point> poly;
+      if (!setPolygonShape(polygon_points, poly)) {
+        return false;
+      }
+
+      // linear_max param
+      nav2_util::declare_parameter_if_not_declared(
```

Comment:
Why can't the VelocityPolygon object get  its own parameters? That would be better separation of concerns

Reply:
I was thinking to validate all the parameters before creating the `VelocityPolygon` object. I think this is do-able, will update to pass the `node_` pointer and handle it's own parameter in the `VelocityPolygon` object.